### ISSUE TITLE
[BUG] Crash Report in firebase

### ISSUE THREAD

**Issue by fmacherey** [Usercentrics/react-native-sdk#26]
**Describe the bug**
Hey guys, we get some crash reports in firebase for the cmp dialog. I know this is very shady, as I don't have a reproducable example, but maybe the stack trace can help:
![image](https://user-images.githubusercontent.com/72257175/152785311-093de02e-29a3-4afa-ac06-98ca59a53248.png)

**usercentrics/react-native branch/commit**
`"@usercentrics/react-native-sdk": "^2.0.4-rc1"`

**Comment by fmacherey** [comment]
I also get this stacktrace:
![image](https://user-images.githubusercontent.com/72257175/152786025-36ea4a77-97b5-4dec-9f10-0d5262e318aa.png)

**Comment by DominikHinc** [comment]
Hey!
We have the same issue in our app, seems like it's Android-only issue.
![image](https://user-images.githubusercontent.com/46762501/153841138-df84c72d-e3e5-4025-8659-10850ea35074.png)
![image](https://user-images.githubusercontent.com/46762501/153841209-579baaca-fa91-4398-87bd-effb4fc5f306.png)
![image](https://user-images.githubusercontent.com/46762501/153841696-5ea9cd63-0872-4923-b943-f1227ae98019.png)
Unfortunately, I also don't know how to reproduce this.

**Comment by uc-pedro** [comment]
We've made some adjustments to the mentioned methods on these stacktraces, it's available on the new 2.2.1 version. 
Make sure that if you update to the new version to read our docs because the showCMP method is now deprecated in favor of showFirstLayer/showSecondLayer methods. 

https://docs.usercentrics.com/cmp_in_app_sdk/latest/collect_consent/present_cmp/

**Comment by adeem** [comment]
We upgraded to latest version of the library and we are still getting lot of crashes from Firebase. Any help?

```
Crashed: com.apple.main-thread
0  Usercentrics                   0x4908e8 ReadHeapRefLocked + 36
1  Usercentrics                   0x13447c kfun:kotlinx.coroutines#<get-mainThread>(){}kotlinx.coroutines.MainThread + 144
2  Usercentrics                   0x3dcfe8 __unnamed_2914 + 524144
3  Usercentrics                   0x492fd0 Kotlin_initRuntimeIfNeeded + 428
4  Usercentrics                   0x488214 +[KotlinBase allocWithZone:] + 36
5  Raisin                         0xe54a14 specialized UsercentricsUsercentricsOptions.init(from:) + 546072 (<compiler-generated>:546072)
6  Raisin                         0xe454ec closure #1 in RNUsercentricsModule.configure(_:) + 483312 (<compiler-generated>:483312)
7  Raisin                         0x106648 thunk for @escaping @callee_guaranteed () -> () + 4371850824 (<compiler-generated>:4371850824)
8  libdispatch.dylib              0x63094 _dispatch_call_block_and_release + 24
9  libdispatch.dylib              0x64094 _dispatch_client_callout + 16
10 libdispatch.dylib              0x10d40 _dispatch_main_queue_drain + 928
11 libdispatch.dylib              0x10990 _dispatch_main_queue_callback_4CF$VARIANT$mp + 36
12 CoreFoundation                 0x4dab4 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
13 CoreFoundation                 0xafd8 __CFRunLoopRun + 2544
14 CoreFoundation                 0x1dc30 CFRunLoopRunSpecific + 572
15 GraphicsServices               0x1988 GSEventRunModal + 160
16 UIKitCore                      0x4e5c50 -[UIApplication _run] + 1080
17 UIKitCore                      0x27f3d0 UIApplicationMain + 336
18 Raisin                         0x598c main + 14 (main.m:14)
19 ???                            0x1039783d0 (Missing)
```

![image](https://user-images.githubusercontent.com/68442/162881944-d945013e-7758-496f-8fe0-409a2a29c78a.png)

**Comment by adeem** [comment]
I am seeing 4 crashes coming because of UserCentric, disturbing lot of our user base. As I can not reproduce this issue, I am also blind on this point. 

1st Crash
![Raisin_–__compiler-generated__line_546072_–_Firebase_console](https://user-images.githubusercontent.com/68442/162984028-ebf740aa-063c-4914-ae4a-eecaa22b2e5b.png)

2nd Crash
![Raisin_–__compiler-generated__line_544552_–_Firebase_console](https://user-images.githubusercontent.com/68442/162984396-2cac9794-5665-4e45-be1f-41e1e4df3ec4.png)

3rd Crash
![Raisin_–__compiler-generated__line_524240_–_Firebase_console](https://user-images.githubusercontent.com/68442/162984535-268e604e-05bb-4417-bd33-ba097c730d67.png)

4th Crash
and crash point is > `9  Usercentrics                   0x4975f4 (anonymous namespace)::TerminateHandler::queuedHandler()::'lambda'()::operator()() const + 22`
```
Crashed: com.google.firebase.crashlytics.ios.exception
0  App                         0xa07e4c FIRCLSProcessRecordAllThreads + 393 (FIRCLSProcess.c:393)
1  App                         0xa0822c FIRCLSProcessRecordAllThreads + 424 (FIRCLSProcess.c:424)
2  App                         0x9ff618 FIRCLSHandler + 34 (FIRCLSHandler.m:34)
3  App                         0x9f9d98 __FIRCLSExceptionRecord_block_invoke + 215 (FIRCLSException.mm:215)
4  libdispatch.dylib              0x41c8 _dispatch_client_callout + 20
5  libdispatch.dylib              0x13414 _dispatch_lane_barrier_sync_invoke_and_complete + 56
6  App                         0x9f9538 FIRCLSExceptionRecord + 219 (FIRCLSException.mm:219)
7  App                         0x9f9840 FIRCLSExceptionRecordNSException + 109 (FIRCLSException.mm:109)
8  App                         0x9f9138 FIRCLSTerminateHandler() + 273 (FIRCLSException.mm:273)
9  Usercentrics                   0x4975f4 (anonymous namespace)::TerminateHandler::queuedHandler()::'lambda'()::operator()() const + 22
10 Usercentrics                   0x4975c4 void (anonymous namespace)::$_0::operator()<(anonymous namespace)::TerminateHandler::queuedHandler()::'lambda'()>((anonymous namespace)::TerminateHandler::queuedHandler()::'lambda'()) + 40
11 Usercentrics                   0x49759c (anonymous namespace)::TerminateHandler::queuedHandler() + 10
12 Usercentrics                   0x497330 (anonymous namespace)::TerminateHandler::kotlinHandler() + 132
13 libc++abi.dylib                0x10ea4 std::__terminate(void (*)()) + 20
14 libc++abi.dylib                0x13ebc __cxa_rethrow + 148
15 libobjc.A.dylib                0x2dcb0 objc_exception_rethrow + 44
16 AppKit                         0x94b1c -[NSWindow _setFrameCommon:display:fromServer:] + 3896
17 SafariPlatformSupport          0x3f08 (Missing UUID b17cf8bdd94339779f0e7d20cc0ecd7d)
18 SafariPlatformSupport          0x3d68 (Missing UUID b17cf8bdd94339779f0e7d20cc0ecd7d)
19 UIKitCore                      0xb2f8e8 -[UIKeyboardImpl generateAutofillCandidateByAddingTask:] + 944
20 UIKitCore                      0xb39af8 -[UIKeyboardImpl setDelegate:force:] + 6380
21 UIKitCore                      0x819ec8 -[UIKeyboardSceneDelegate _reloadInputViewsForKeyWindowSceneResponder:] + 2024
22 UIKitCore                      0x8196b4 -[UIKeyboardSceneDelegate _reloadInputViewsForResponder:] + 164
23 UIKitCore                      0x18a1f0 -[UIResponder(UIResponderInputViewAdditions) reloadInputViews] + 132
24 UIKitCore                      0x132750 -[UIResponder becomeFirstResponder] + 856
25 UIKit                          0xc0bec -[UITextInputUIResponderAccessibility becomeFirstResponder] + 52
26 UIKitCore                      0x1625ac -[UIView(Hierarchy) becomeFirstResponder] + 148
27 WebKit                         0x7d0548 -[WKContentView(WKInteraction) becomeFirstResponderForWebView] + 152
28 WebKit                         0x31fb6c -[WKWebView(WKViewInternalIOS) becomeFirstResponder] + 148
29 WebKit                         0x7e377c -[WKContentView(WKInteraction) _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:] + 1864
30 WebKit                         0x3fc278 WebKit::WebPageProxy::elementDidFocus(WebKit::FocusedElementInformation const&, bool, bool, WTF::OptionSet<WebCore::ActivityState::Flag>, WebKit::UserData const&) + 216
31 WebKit                         0x7482b4 WebKit::WebPageProxy::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 62776
32 WebKit                         0x31ac98 IPC::MessageReceiverMap::dispatchMessage(IPC::Connection&, IPC::Decoder&) + 264
33 WebKit                         0x4e59b8 WebKit::WebProcessProxy::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 40
34 WebKit                         0x3153bc IPC::Connection::dispatchMessage(std::__1::unique_ptr<IPC::Decoder, std::__1::default_delete<IPC::Decoder> >) + 788
35 WebKit                         0x314a70 IPC::Connection::dispatchIncomingMessages() + 508
36 JavaScriptCore                 0x11bfb8c WTF::RunLoop::performWork() + 200
37 JavaScriptCore                 0x11c0928 WTF::RunLoop::performWork(void*) + 36
38 CoreFoundation                 0x849a0 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28
39 CoreFoundation                 0x848ec __CFRunLoopDoSource0 + 208
40 CoreFoundation                 0x845ec __CFRunLoopDoSources0 + 268
41 CoreFoundation                 0x82f6c __CFRunLoopRun + 828
42 CoreFoundation                 0x82490 CFRunLoopRunSpecific + 600
43 HIToolbox                      0x33dd8 RunCurrentEventLoopInMode + 292
44 HIToolbox                      0x33b54 ReceiveNextEventCommon + 564
45 HIToolbox                      0x33908 _BlockUntilNextEventMatchingListInModeWithFilter + 72
46 AppKit                         0x41bcc _DPSNextEvent + 860
47 AppKit                         0x404c4 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1328
48 AppKit                         0x32690 -[NSApplication run] + 596
49 AppKit                         0x3da8 NSApplicationMain + 1132
50 AppKit                         0x2d9e70 _NSApplicationMainWithInfoDictionary + 22
51 UIKitMacHelper                 0x4a7c UINSApplicationMain + 1276
52 UIKitCore                      0x39f8 UIApplicationMain + 164
53 App                         0x598c main + 14 (main.m:14)
54 ???                            0x105b5d088 (Missing)
```

**Comment by adeem** [comment]
As you can see from firebase crash reporting, those crashes happened at the start of the app (within the first 2 seconds). And we were using the past 2.2.1 version of user-centric and now migrated to the latest version of usercentric. Is there any migration issue?
Any other guess you might have for so many crashes?

**Comment by uc-pedro** [comment]
@adeem do you have any multithreading manager that you might use within Usercentrics, or are you calling Usercentrics from any other thread different than the main one? 

Can you give us some info about where you call the configure method in your app lifecycle?

Also, we are about to release a new version with some fixes in the initialization flow, but we don't know for sure if it will fix the issue that you're facing since we could not reproduce it yet.

**Comment by adeem** [comment]
No, we are calling Usercentrics normally from our codebase without any multithreading manager. (https://reactnative.dev/docs/native-modules-ios#threading)

InteractionManager might cause problems for usercentric?
![image (7)](https://user-images.githubusercontent.com/68442/164195916-79f6f0aa-2006-439c-ad3f-96166316653b.png)

**Comment by adeem** [comment]
any update @uc-pedro ? 
When app is launched, we do the onboarding for the customer and once they select the country, we show them user centric modal view.

**Comment by adeem** [comment]
Accidently I was able to crash the app. Here how it looks like. SettingID was not null, I was able to print its value. 



![Cursor_and_Raisin_—_UsercentricsOptions_Dict_swift](https://user-images.githubusercontent.com/68442/164885653-e93e12e6-58af-4444-b5d6-2a6e64783ff3.png)

**Comment by uc-noack** [comment]
Hi @adeem,
Thanks for the insights. We will look into it.
Have you managed to reproduce this issue consistantly? Was there something specific you did? or a specific part of your user journey that triggered this issue?

**Comment by uc-leo** [comment]
hey @adeem, I was taking a look into the thread and also into your latest informations provided (btw, thanks for that!), and I was able to reproduce the exact problem using the code you sent.

so I did a deeper investigation and also took the time to download your app to perform some manual testing... I tested on Android and iOS and didn't received a single crash 🙌 (not sure why some users are getting those errors...)

going to the point you raised in your last commend:
> how to configure the usercentric with await?
 
you don't need to do that, since it's already an async method which will not return anything, you'll need to subscribe to `Usercentrics.status()` (like you're doing) to get its result.

in your case I saw that after the Onboarding, you show the Country selection screen and after that you actually display the Usercentrics Popup (probably because there are specifics SettingsIDs for each country), it's indeed not a common scenario, but that's not a problem at all.

the reason that the code works normally when you add `await delay(1000);` is because that routine is dispatched to run in a new thread. the `configure` method dispatches its routine to run in the Main Thread, therefore, when the code blocks inside `useEffect()` runs, they are running in different threads and the queued `configure` method still's waiting it's turn to be executed.

in order to fix it, we think that'll be a better idea to release a new version adjusting this behaviour! this will happen in the next days, I'll let you know when you can test it again, sounds good?

thanks again for your patience and all the details! 😄

**Comment by adeem** [comment]
In our latest release we added a delay and that stopped the crash :) 

so we are waiting for your update ❤️

**Comment by MegaMaddin** [comment]
Hey,

seeing the same issue as @adeem, lots of crashes with a similar signature:

<img width="1240" alt="image" src="https://user-images.githubusercontent.com/4417135/165564989-3ca69dd0-cc9e-457a-b456-ac2ca7da50c6.png">

> in order to fix it, we think that'll be a better idea to release a new version adjusting this behaviour! this will happen in the next days, I'll let you know when you can test it again, sounds good?

Is there any update on that new version? I was using the latest available version (2.3.0).

On the issue itself, wouldn't it make sense to either make the `isReady` waiting and blocking or erroring out when the SDK isn't initialized or make the `configure` awaitable?

**Comment by MegaMaddin** [comment]
Sorry, just saw that this update is just from yesterday. I've disabled the SDK for now and waiting for a new release.

**Comment by uc-leo** [comment]
@adeem @MegaMaddin 

the new release 2.4.0 is out! it should fix this problem 🎉 

let me know if you need anything else!

thanks! 🙏

**Comment by uc-leo** [comment]
@adeem @MegaMaddin

the new release 2.4.0 is out! it should fix this problem 🎉

let me know if you need anything else!

thanks!

**Comment by MegaMaddin** [comment]
Currently building a release candidate with the latest fixes. Fingers crossed!

**Comment by MegaMaddin** [comment]
Sorry for typos, on the phone. It looks like this introduced another issue, if the user was on a previous version of the SDK. I tested the upgrade path from 2.3.0 to 2.4.0, and as long as the user does not "store" once with the new version of the SDK, the app crashes randomly on start. The crash seems to happen so early, that I even don't see any reports in Crashlytics. Once I store (agree or deny the consent) once with the new SDK, the crash seems to be gone (so eventually a on disk format issue, or data serialization issue?). Will further investigate tomorrow and see if I can get crash data in Instruments. @uc-pedro @uc-leo

**Comment by MegaMaddin** [comment]
Just FYI - I released SDK 2.4.0 yesterday and so far I can't see any crashes on iOS/Android. @uc-leo @uc-pedro

**Comment by adeem** [comment]
@MegaMaddin  I didn't understand your message. You were suggesting before that the app was crashing and now it's not? I thought you were using the 2.4.0 (which was released 15 days ago).
I was waiting for your issue to be resolved before I upgrade the library on our side.

**Comment by MegaMaddin** [comment]
@adeem Yes and no. I was still seeing crashes on my physical test devices, but w/o any traces (nothing in Crashlytics, nothing in  crashlogs etc. etc.). So I can't be 100% confident that those where related to the SDK. On the other hand that was the only change I made at that time. Since I have a tight schedule wrt CMP and needed to go forward with this we made the decision to release that version of the SDK anyway. So far we haven't got any bad feedback from users and I can't see any crashes in Crashlytics (or maybe they just not appear there as with my test devices 🤷🏼‍♂️). Rolling back and making another release w/o the SDK is easier than waiting for a solution from a third party.
